socks/socks.go

```diff
@@ -163,29 +164,65 @@ func ParseAddr(s string) Addr {
 
 // Handshake fast-tracks SOCKS initialization to get target address to connect.
 func Handshake(rw io.ReadWriter) (Addr, error) {
-	// Read RFC 1928 section 4 for request and reply structure and sizes
-	buf := make([]byte, MaxReqLen)
+	// Read RFC 1928 for request and reply structure and sizes
 
-	_, err := rw.Read(buf) // SOCKS version and auth methods
-	if err != nil {
+	buf := make([]byte, MaxReqLen)
+	// read VER, NMETHODS, METHODS
+	if _, err := io.ReadFull(rw, buf[:2]); err != nil {
 		return nil, err
 	}
-
-	_, err = rw.Write([]byte{5, 0}) // SOCKS v5, no auth required
-	if err != nil {
+	if buf[0] != 5 { // VER: 5
```

Comment:
No need to be very strict about SOCKS version. SOCKS4 is about the same. In general the program should be less picky about its input.